tests/net_dynamic_hb.rs

```diff
@@ -471,6 +479,73 @@ where
             net,
             join_plan: None,
             shutdown_epoch: None,
+            saved_nodes: Vec::new(),
+            old_subset_applied: false,
         }
     }
+
+    /// Selects random subset of validators which can be safely removed from the network
+    ///
+    /// The cluster always remain correct after removing this subset from the custer.
+    /// This method may select as correct nodes as malicious.
+    fn subset_for_remove<R>(&self, rng: &mut R) -> BTreeSet<usize>
+    where
+        R: rand::Rng,
+    {
+        let net = &self.net;
+        let (faulty, correct): (Vec<_>, Vec<_>) = net.nodes().partition(|n| n.is_faulty());
+
+        let f = faulty.len();
+        let n = correct.len() + f;
+
+        assert!(
+            n > f * 3,
+            "the network is already captured by the faulty nodes"
+        );
+
+        let max_number_correct_nodes_for_remove = n - (f * 3 + 1);
+        let remove_from_correct = rng.gen_range(0, max_number_correct_nodes_for_remove + 1);
+        let remove_from_faulty = if remove_from_correct == 0 && f > 0 {
+            // if we can't remove any correct node, because network will become in an
+            // incorrect state we have to remove at least 1 faulty node
```

Comment:
I think we only actually remove the nodes from the network after all correct nodes got `Complete(NodeChange(new_pub_keys))`, do we?
In that case, we shouldn't have to care about the _old_ `n`: We just need to make sure that less than a third of the _new_ set of nodes is faulty.

E.g. if `n` is 7 and `f` is 2, we should be able to remove two correct and one faulty node, because then `new_n` is 5 and `new_f` is 1, which is still less than a third. Even though 2 faulty out of 5 would be too much.

Reply:
`do we?` - sure! 
`We just need to make sure that less than a third of the _new_ set of nodes is faulty.` - yes. Old set is always correct.
`E.g. if n is 7 and f is 2...` - I agree. Current implementation selects only 1 or 2 malicious nodes in this case. Do you want to improve `subset_for_remove` that it will be able to select 2 correct and 1 malicious node as well?

Reply:
I'd say so, yes. Maybe just pick a random `new_n < n` first, and then remove as many correct nodes as possible and as many faulty nodes as necessary to keep the new set >⅔ correct.

Reply:
done